File: 2023/nim/p05.nim
import strutils, sequtils, algorithm, sugar

let
    input = readFile("../data/05.txt").strip.splitLines
    seeds = input[0].split[1..^1].mapIt(it.strip.parseInt)

func normalizeRange(ranges: seq[(int, int, int)]): auto =
    var res: seq[(int, int)] = @[]
    var prev = -1
    for item in ranges.sorted:
        if prev >= 0 and prev != item[0]:
            res.add((prev, 0))
        res.add((item[0], item[2]))
        prev = item[1]
    res.add((prev, 0))
    res

func parseMap(input: seq[string]): auto =
    var maps: seq[seq[(int, int)]] = @[]
    var ranges: seq[(int, int, int)] = @[]
    for line in input[3..^1]:
        if not line.isEmptyOrWhitespace:
            if line[^1] == ':':
                maps.add(normalizeRange(ranges))
                ranges = @[]
            else:
                let
                    nums = line.split.mapIt(it.parseInt)
                    x = nums[0]
                    y = nums[1]
                    z = nums[2]
                ranges.add (y, y+z, x-y)
    maps.add(normalizeRange(ranges))
    maps

func findMapIndex(mappings: seq[(int, int)], target: int): int =
    for i, mp in mappings:
        if mp[0] > target:
            return i-1
    return mappings.len-1

func convert(source: int, mappings: seq[(int, int)]): int =
    let i = findMapIndex(mappings, source)
    source + mappings[i][1]

func convertRange(srcRanges: seq[(int, int)], mappings: seq[(int, int)]): seq[(
        int, int)] =
    var res: seq[(int, int)] = @[]
    for src in srcRanges:
        let
            (fro, til) = src
            i = findMapIndex(mappings, fro)
            j = findMapIndex(mappings, til-1)
        var diff = mappings[i][1]
        if i == j:
            res.add((fro + diff, til + diff))
        else:
            res.add((fro + diff, mappings[i+1][0] + diff))
            for k in (i+1..j-1):
                diff = mappings[k][1]
                res.add((mappings[k][0] + diff, mappings[k+1][0] + diff))
            diff = mappings[j][1]
            res.add((mappings[j][0] + diff, til + diff))
    res

let maps = parseMap(input)

# part 1
echo seeds.map(seed => maps.foldl(convert(a, b), seed)).min

# part 2
var seedRanges: seq[(int, int)] = @[]
for i in (0..(seeds.len div 2 - 1)):
    let x = seeds[2*i]
    let y = seeds[2*i+1]
    seedRanges.add((x, x+y))
echo maps.foldl(convertRange(a, b), seedRanges).min[0]

File: 2023/ruby/p05.rb
input = IO.foreach('../data/05.txt').map(&:strip)

seeds = input[0].split[1..-1].map(&:to_i)

# sort and fill the gaps with 0-diff ranges
def normalize_range(ranges)
  res = []
  prev = nil
  ranges.sort.each{|from, to, diff|
    res << [prev, 0] if prev && from != prev
    res << [from, diff]
    prev = to
  }
  res << [prev, 0]
end

def parse_map(input)
  maps = []
  ranges = []
  input[3..-1].each{|line|
    next if line.empty?
    if line[-1] == ':'
      maps << normalize_range(ranges)
      ranges = []
    else
      x,y,z = line.split.map(&:to_i)
      ranges << [y, y+z, x-y] # [src start, src end, diff from src to dest]
    end
  }
  maps << normalize_range(ranges)
end

maps = parse_map(input)

def find_map_index(mappings, target)
  i = mappings.bsearch_index{|x| x[0] > target} || mappings.size
  i-1
end

def convert(source, mappings)
  i = find_map_index(mappings, source)
  source + mappings[i][1]
end

def convert_range(src_ranges, mappings)
  src_ranges.flat_map{|from, til|
    i = find_map_index(mappings, from)
    j = find_map_index(mappings, til-1)
    if i == j
      diff = mappings[i][1]
      [[from + diff, til + diff]]
    else
      nxt = []
      diff = mappings[i][1]
      nxt << [from + diff, mappings[i+1][0] + diff]
      (i+1).upto(j-1){|k|
        diff = mappings[k][1]
        nxt << [mappings[k][0] + diff, mappings[k+1][0] + diff]
      }
      diff = mappings[j][1]
      nxt << [mappings[j][0] + diff, til + diff]
      nxt
    end
  }
end

# part 1
p seeds.map{|seed| maps.reduce(seed){|src,mp| convert(src, mp)}}.min

# part 2
seed_ranges = seeds.each_slice(2).map{|x, y| [x, x + y]}
p maps.reduce(seed_ranges){|src, mp| convert_range(src, mp)}.min[0]
